### PR TITLE
Add option to remove a canister with canister_ids tool

### PR DESCRIPTION
**Pull request by dskloetd**
# Motivation

Trying to make the release a less manual process.
One of the steps in the release process is to remove an entry from `canister_ids.json` to deploy a new canister to `staging`.

# Changes

Update `scripts/canister_id` to:
1. Use `clap`.
2. Properly pass args to `jq` instead of interpolating in the filter.
3. Add functionality to remove a canister ID (it leaves it with a backup name just in case).

# Tests

1. Added `scripts/canister_ids.test` to test existing and new functionality.
2. Added the test to GitHub CI.

# Todos

- [x] Add entry to changelog (if necessary).

### CODE REVIEW

**Comment by bitdivine** [comment]
Regarding the changelog conflict, if you merge in main locally, git will probably resolve the conflict for you.  GitHub doesn't pay attention to the merge rules.

File: scripts/canister_ids [hunk 1]
```diff
@@ -1,37 +1,70 @@
...
-ids_file="$top_dir/canister_ids.json"
+# Source the clap.bash file ---------------------------------------------------
+source "$SOURCE_DIR/clap.bash"
+# Define options
+clap.define short=e long=export desc="Export canister IDs" variable=EXPORT nargs=0 default="false"
+clap.define short=i long=import desc="Import canister IDs from stdin" variable=IMPORT nargs=0 default="false"
+clap.define short=r long=remove desc="Remove canister ID" variable=REMOVE nargs=0 default="false"
+clap.define short=n long=network desc="dfx network to use" variable=NETWORK default="staging"
+clap.define short=c long=canister desc="canister to remove" variable=CANISTER default="nns-dapp"
+clap.define short=j long=json-file desc="Path to the canister_ids.json file" variable=JSON_FILE default="./canister_ids.json"
```
[the remark this answers is not in the document]
Comment (dskloetd) [reply scripts/canister_ids:25]: The only use case I had in mind is the one I described above.
The only reason I added a flag for the file is to be able to test the script without destroying one's main `canister_ids.json`.

I'm happy to consider other use cases but I currently don't really understand them.

File: scripts/canister_ids [hunk 2]
```diff
@@ -1,37 +1,70 @@
...
+fi
+
+if ! [ "${command:-}" ]; then
+  print_help
+  exit 0
+fi
 
+if [ "$command" = "export" ]; then
   # This whole block can be done in a single line of jq script but it woudn't
   # be readable.
```
[the remark this answers is not in the document]
Comment (dskloetd) [reply scripts/canister_ids:61]: I know what it should do and I still can't read that. But I'll change it if you can find one more person on the nss-dapp team who finds the `jq` filter easier to read.

File: scripts/canister_ids [hunk 3]
```diff
@@ -42,9 +75,26 @@ fi
...
+
+  if ! [ "$old_canister_id" ]; then
+    echo "$CANISTER already didn't have an ID for $NETWORK."
+    exit
+  fi
+
+  new_name="$NETWORK-before-$(date +%Y%m%d-%H%M%S)"
+  tmp_file="$(mktemp)"
+  jq --arg network "$NETWORK" --arg canister "$CANISTER" --arg new_name "$new_name" '.[$canister][$new_name] = .[$canister][$network] | del(.[$canister][$network])' "$JSON_FILE" >"$tmp_file"
+  mv "$tmp_file" "$JSON_FILE"
```
[the remark this answers is not in the document]
Comment (dskloetd) [reply scripts/canister_ids]: I wasn't sure if `sponge` is installed everywhere, but I've changed it.

File: scripts/canister_ids.test [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  "__Candid_UI": {
+    "staging": "tmxop-wyaaa-aaaaa-aaapa-cai"
+  },
+  "internet_identity": {
+    "staging": "wflfh-4yaaa-aaaaa-aaata-cai"
+  },
+  "nns-dapp": {
+    "mainnet": "qoctq-giaaa-aaaaa-aaaea-cai",
+    "staging": "4yr67-tiaaa-aaaaa-aabka-cai"
+  }
```
[the remark this answers is not in the document]
Comment (dskloetd) [reply scripts/canister_ids.test:22]: Done

The `export` is unaffected but the `import` and `remove` tests keep the entry.

**Comment by dskloetd** [comment]
> What happens if a user imports canisters for a given network but there are existing entries for that network?

They overwritten. It's covered by the test.

File: scripts/canister_ids (hunk 3 above)
[the remark this answers is not in the document]
Comment (dskloetd) [reply scripts/canister_ids]: Now it's failing on CI:
https://github.com/dfinity/nns-dapp/actions/runs/5403416693/jobs/9816200091?pr=2768
> /home/runner/work/nns-dapp/nns-dapp/scripts/canister_ids: line 94: sponge: command not found

Is it really worth installing more software? Or shall I change it back?

**Review by bitdivine — APPROVED** [review]

File: scripts/canister_ids (hunk 1 above)
[the remark this answers is not in the document]
Comment (dskloetd) [reply scripts/canister_ids:25]: Let's add the local use case later if it looks like it is useful.

File: scripts/canister_ids (hunk 2 above)
[the remark this answers is not in the document]
Comment (bitdivine) [reply scripts/canister_ids:61]: Well, in the poll Andronikos opted for the jq and others were ambivalent. :-) But no pressure.

[the remark this answers is not in the document]
Comment (dskloetd) [reply scripts/canister_ids:61]: I asked him to explain the `jq` command to me and he struggled a lot. He didn't actually try to really understand either code before voting so I don't think he answered the question of what's easier to read for him. I didn't ask any of the other people. But if you send me a PR, I won't block it.

**Pull request merged**